File: Android/SDK/liveroom/src/main/java/com/aliyun/standard/liveroom/lib/component/view/LiveLoadingView.java
package com.aliyun.standard.liveroom.lib.component.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.ContentLoadingProgressBar;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import com.aliyun.roompaas.live.LiveEvent;
import com.aliyun.roompaas.live.SampleLiveEventHandler;
import com.aliyun.standard.liveroom.lib.LiveContext;
import com.aliyun.standard.liveroom.lib.LivePrototype;
import com.aliyun.standard.liveroom.lib.R;
import com.aliyun.standard.liveroom.lib.component.BaseComponent;
import com.aliyun.standard.liveroom.lib.component.ComponentHolder;
import com.aliyun.standard.liveroom.lib.component.IComponent;

/**
 * @author puke
 * @version 2022/3/21
 */
public class LiveLoadingView extends FrameLayout implements ComponentHolder {

    public static final String ACTION_SHOW_LOADING = "show_loading";
    public static final String ACTION_HIDE_LOADING = "hide_loading";

    protected final ContentLoadingProgressBar loadingBar;

    private final Component component = new Component();
    private final boolean supportLoadingView;

    public LiveLoadingView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        supportLoadingView = LivePrototype.getInstance().getOpenLiveParam().supportLoadingView;
        setVisibility(GONE);
        inflate(context, R.layout.ilr_view_live_loading, this);
        loadingBar = findViewById(R.id.loading_bar);
    }

    protected void startLoading() {
        if (supportLoadingView) {
            setVisibility(VISIBLE);
            loadingBar.show();
        }
    }

    protected void endLoading() {
        setVisibility(GONE);
        loadingBar.hide();
    }

    @Override
    public IComponent getComponent() {
        return component;
    }

    private class Component extends BaseComponent {
        @Override
        public void onInit(LiveContext liveContext) {
            super.onInit(liveContext);
            if (!supportLoadingView) {
                return;
            }

            liveService.addEventHandler(new SampleLiveEventHandler() {
                @Override
                public void onLoadingBegin() {
                    startLoading();
                }

                @Override
                public void onLoadingEnd() {
                    endLoading();
                }

                @Override
                public void onPrepared() {
                    endLoading();
                }

                @Override
                public void onPusherEvent(LiveEvent event) {
                    switch (event) {
                        case RECONNECT_START:
                            startLoading();
                            break;
                        // 重新推流成功时, 会回调FIRST_FRAME_PUSHED
                        case FIRST_FRAME_PUSHED:
                        case RECONNECT_SUCCESS:
                        case RECONNECT_FAIL:
                        case CONNECTION_FAIL:
                            endLoading();
                            break;
                    }
                }
            });
        }

        @Override
        public void onEvent(String action, Object... args) {
            switch (action) {
                case ACTION_SHOW_LOADING:
                    startLoading();
                    break;
                case ACTION_HIDE_LOADING:
                    endLoading();
                    break;
            }
        }
    }
}

File: Android/Demo/liveroom-demo/ui-enterprise/src/main/java/com/aliyun/vpaas/standard/enterprise/view/LiveLoadingView.kt
package com.aliyun.vpaas.standard.enterprise.view

import android.content.Context
import android.support.v4.widget.ContentLoadingProgressBar
import android.util.AttributeSet
import android.widget.FrameLayout
import com.aliyun.player.bean.ErrorInfo
import com.aliyun.roompaas.live.SampleLiveEventHandler
import com.aliyun.roompaas.live.exposable.event.LiveCommonEvent
import com.aliyun.standard.liveroom.lib.Actions
import com.aliyun.standard.liveroom.lib.LiveContext
import com.aliyun.standard.liveroom.lib.LivePrototype
import com.aliyun.standard.liveroom.lib.component.BaseComponent
import com.aliyun.standard.liveroom.lib.component.ComponentHolder
import com.aliyun.standard.liveroom.lib.component.IComponent
import com.aliyun.standard.liveroom.lib.component.view.LiveLoadingView
import com.aliyun.vpaas.standard.enterprise.R
import java.text.SimpleDateFormat
import java.util.*

/**
 * @author puke
 * @version 2022/3/21
 */
class LiveLoadingView(context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs),
    ComponentHolder {

    companion object {
        const val ACTION_SHOW_LOADING = "show_loading"
        const val ACTION_HIDE_LOADING = "hide_loading"
    }

    protected val loadingBar: ContentLoadingProgressBar
    private val component: Component = Component()
    private val supportLoadingView = LivePrototype.getInstance().openLiveParam.supportLoadingView

    init {
        visibility = GONE
        inflate(context, R.layout.ep_view_live_loading, this)
        loadingBar = findViewById(R.id.loading_bar)
    }

    private fun showLoading() {
        if (supportLoadingView) {
            visibility = VISIBLE
        }
    }

    private fun hideLoading() {
        visibility = GONE
    }

    override fun getComponent(): IComponent {
        return component
    }

    private inner class Component : BaseComponent() {
        override fun onInit(liveContext: LiveContext) {
            super.onInit(liveContext)
            if (!supportLoadingView) {
                return
            }
            liveService.addEventHandler(object : SampleLiveEventHandler() {
                override fun onLiveStarted(event: LiveCommonEvent?) {
                    // 开始直播时, 显示加载框 (刚进入的时候没开播, 在页面内部时开播了)
                    showLoading()
                }

                override fun onLoadingBegin() {
                    showLoading()
                }

                override fun onLoadingEnd() {
                    hideLoading()
                }

                override fun onPrepared() {
                    hideLoading()
                }

                override fun onPlayerError(errorInfo: ErrorInfo?) {
                    hideLoading()
                }
            })
        }

        override fun onEvent(action: String?, vararg args: Any?) {
            when (action) {
                ACTION_SHOW_LOADING -> showLoading()
                ACTION_HIDE_LOADING -> hideLoading()
                Actions.GET_LIVE_DETAIL_SUCCESS -> {
                    liveService?.liveDetail?.liveInfo?.let {
                        val status = liveService?.liveDetail?.liveInfo?.status
                        if (status == LiveStateView.STATUS_LIVING) {
                            // 直播正在进行 (刚进入的时候已经开播了)
                            showLoading()
                        }
                    }
                }
            }
        }
    }
}